core/src/main/java/org/infinispan/persistence/sifs/Compactor.java

```diff
@@ -240,15 +239,9 @@ private void recordFreeSpace(Stats stats, int file, int size) {
    }
 
    private void schedule(int file, Stats stats) {
-      boolean shouldSchedule = false;
-      synchronized (stats) {
-         if (!stats.isScheduled()) {
-            log.debugf("Scheduling file %d for compaction: %d/%d free", file, stats.free.get(), stats.total);
-            stats.setScheduled();
-            shouldSchedule = true;
-         }
-      }
-      if (!terminateSignal && shouldSchedule) {
+      assert stats.isScheduled();
```

Comment:
This is really the key here, instead of allowing anyone to call schedule the caller must first mark the stats as scheduled first in an atomic way to know they can indeed schedule it.